### ISSUE TITLE
Using pretrained weights on REFUGE dataset does not give good results

### ISSUE THREAD

**Issue by jacoblam3112** [MedicineToken/Medical-SAM2#9]
Hi, Thank you for sharing your code. I downloaded the pretrained weights (MedSAM2_pretrain.pth) from the link you provided and loaded them in the model and run the evaluation via file train_2d.py on the REFUGE dataset (also downloaded from your provided link) without any finetuning. I was hoping the pretrained model would give decent results so I could make sure the inference pipeline works ok. But I see very low numbers as give below:

INFO:root:Total score: 1.5592443943023682, IOU: 0.01751479435546341, DICE: 0.029827685931491824 || @ epoch 0.                          
Total score: 1.5592443943023682, IOU: 0.01751479435546341, DICE: 0.029827685931491824 || @ epoch 0.

For your reference, I slightly modify the code in train_2d.py to directly do the validation instead of training first epoch. The rest of the arguments used are as below:

Namespace(b=1, data_path='./dataset/REFUGE', dataset='REFUGE', distributed='none', encoder='vit_b', exp_name='REFUGE_MedSAM2', gpu=True, gpu_device=0, image_size=1024, lr=0.0001, memory_bank_size=16, multimask_output=1, net='sam2', out_size=1024, path_helper={'prefix': 'logs/REFUGE_MedSAM2_2024_08_13_07_04_47', 'ckpt_path': 'logs/REFUGE_MedSAM2_2024_08_13_07_04_47/Model', 'log_path': 'logs/REFUGE_MedSAM2_2024_08_13_07_04_47/Log', 'sample_path': 'logs/REFUGE_MedSAM2_2024_08_13_07_04_47/Samples'}, pretrain='MedSAM2_pretrain.pth', prompt='bbox', prompt_freq=2, sam_ckpt='./checkpoints/sam2_hiera_tiny.pt', sam_config='sam2_hiera_t', train_vis=False, val_freq=1, video_length=2, vis=True, weights=0)

Could you advise what could be the problem here ?

Thank you

**Comment by BA88** [comment]
Thank you, @jiayuanz3, @YunliQi , @WuJunde for developing and sharing this code.  We are very excited to try to segment RGB images without needing to prompt on every slice.

I have the same issue that @jacoblam3112 reports.  Similar to what they did, i did this:

- used provided pretrained weights
- ran function `validation_sam` only on REFUGE Test dataset
- results are poor -- similar to what @jacoblam3112 listed

Next, I trained my own model using `train_2d.py`.  Then I did this:

- ran function `validation_sam` only on REFUGE Validation dataset
- poor results
- ran function `validation_sam` only on REFUGE Test dataset
- poor results

Do you have code that runs inference on never-seen data, eg, REFUGE Validation dataset, from a model saved as shown in `train_2d.py`:
```
            if  edice > best_dice:
                best_dice = edice
                best_tol = tol
                is_best = True

                save_checkpoint({
                'epoch': epoch + 1,
                'model': args.net,
                'state_dict': sd,
                'optimizer': optimizer.state_dict(),
                'best_dice': best_dice,
                'best_tol': best_tol,
                'path_helper': args.path_helper,
            }, is_best, args.path_helper['ckpt_path'], filename="best_dice_checkpoint.pth")
            else:
                is_best = False
```

I am happy to test and share back to you any code you might provide.  Thank you.

**Comment by WuJunde** [comment]
the pertained weight is for 3d, not 2d...

2d weight would still take a while to release, since it has a bug in saving and loading the weights

@jiayuanz3 is working on this

**Comment by BA88** [comment]
Great, thank you so much for the update!
Best wishes.  Nice job!

**Comment by jiayuanz3** [comment]
Sorry for the late reply. I'm still investigating the weird error in saving and loading weights 😢
For your current work, you might need to run the training with validation directly after each epoch. The visualisation results can be set to save at the same time.
Thanks for pointing this out and I apologise again for any inconvenience brought to you 😞

**Comment by owenip** [comment]
Hey @jiayuanz3 ,
I am also having the problem with saving and loading the SAM2 weight but my fine tuned weights are for original 2D segmentation task, not approach from the Medical-SAM2. 
Does your fine tuned model work fine during training but perform weirly with loading the wieght from pt files?

**Comment by jiayuanz3** [comment]
Yes... I don't know why it happens. But if you evaluate directly after each training epoch, it works fine. I'll update if I have any progress. Thanks for pointing out!

**Comment by owenip** [comment]
That is exactly my situation except I am fine tuning the mask decoder only.
So far I tried only loading the state_dict of mask decoder, saving/loading the entire model with torch. No luck so fat.
I will update here if making any progress.
Thanks.

**Comment by 1275468127** [comment]
> That is exactly my situation except I am fine tuning the mask decoder only. So far I tried only loading the state_dict of mask decoder, saving/loading the entire model with torch. No luck so far. I will update here if making any progress. Thanks.
I am also experiencing an issue with saving and loading the SAM2 weights. Have you resolved it yet?

**Comment by 1275468127** [comment]
Hey @owenip ，I am also experiencing an issue with saving and loading the SAM2 weights. Have you resolved it yet?

**Comment by jiayuanz3** [comment]
Sorry again for the delayed error fixing process and it should work now. For validation, you can use the **function.validation_sam** in **train_2d.py**. Just simply replace the **-sam_ckpt** from SAM2's weight to your pretrained weight 😯

**Comment by yaotingwangofficial** [comment]
Dear all,

May I know the pretrained weights provided in huggingface (https://huggingface.co/jiayuanz3/MedSAM2_pretrain/tree/main) is still for 3D data?  Have I missed any information regarding the availability of pretrained weights for 2D data?  

I used the provided pretrained weights on the REFUGE dataset, but the results are still not satisfactory. Should I train from scratch for 2D data?

Any information would be greatly appreciated!

@jiayuanz3 @WuJunde

**Comment by liuna0420** [comment]
> Sorry again for the delayed error fixing process and it should work now. For validation, you can use the **function.validation_sam** in **train_2d.py**. Just simply replace the **-sam_ckpt** from SAM2's weight to your pretrained weight 😯抱歉再次延迟错误修复过程，它现在应该工作了。为了验证，您可以使用这个函数。train_d_py中的Validation_sam。只需将-sam_ckpt从SAM2的权重替换为预训练的权重😯

I also encountered a related issue. When using the pre-trained model for testing, the results were very poor. However, the validation results during training were quite good. Am I mistakenly using the validation results as the test results?  

Looking forward to your reply!

**Comment by liuna0420** [comment]
> Hi, Thank you for sharing your code. I downloaded the pretrained weights (MedSAM2_pretrain.pth) from the link you provided and loaded them in the model and run the evaluation via file train_2d.py on the REFUGE dataset (also downloaded from your provided link) without any finetuning. I was hoping the pretrained model would give decent results so I could make sure the inference pipeline works ok. But I see very low numbers as give below:嗨，谢谢你分享你的代码。我从您提供的链接中下载了预训练的权重（MedSAM2_pretrain.pth），并将它们加载到模型中，并通过REFUGE数据集（也从您提供的链接中下载）上的文件train_2d.py运行评估，没有进行任何微调。我希望预训练的模型会给出不错的结果，这样我就可以确保推理管道工作正常。但我看到的数字非常低，如下所示：
> 
> INFO:root:Total score: 1.5592443943023682, IOU: 0.01751479435546341, DICE: 0.029827685931491824 || @ epoch 0.INFO：根：总分：1.5592443943023682,IOU: 0.01751479435546341, DICE: 0.029827685931491824 || @ epoch 0。 Total score: 1.5592443943023682, IOU: 0.01751479435546341, DICE: 0.029827685931491824 || @ epoch 0.总分：1.5592443943023682,IOU: 0.01751479435546341, DICE: 0.029827685931491824 || @ epoch 0。
> 
> For your reference, I slightly modify the code in train_2d.py to directly do the validation instead of training first epoch. The rest of the arguments used are as below:为了供您参考，我稍微修改了train__2 .py中的代码，以直接进行验证，而不是训练第一个epoch。使用的其他参数如下：
> 
> Namespace(b=1, data_path='./dataset/REFUGE', dataset='REFUGE', distributed='none', encoder='vit_b', exp_name='REFUGE_MedSAM2', gpu=True, gpu_device=0, image_size=1024, lr=0.0001, memory_bank_size=16, multimask_output=1, net='sam2', out_size=1024, path_helper={'prefix': 'logs/REFUGE_MedSAM2_2024_08_13_07_04_47', 'ckpt_path': 'logs/REFUGE_MedSAM2_2024_08_13_07_04_47/Model', 'log_path': 'logs/REFUGE_MedSAM2_2024_08_13_07_04_47/Log', 'sample_path': 'logs/REFUGE_MedSAM2_2024_08_13_07_04_47/Samples'}, pretrain='MedSAM2_pretrain.pth', prompt='bbox', prompt_freq=2, sam_ckpt='./checkpoints/sam2_hiera_tiny.pt', sam_config='sam2_hiera_t', train_vis=False, val_freq=1, video_length=2, vis=True, weights=0)名称空间(b = 1, data_path = '。/dataset/REFUGE', dataset='REFUGE', distributed='none', encoder='vit_b', exp_name='REFUGE ', gpu=True, gpu_device=0, image_size=1024, lr=0.0001, memory_bank_size=16, multimask_output=1, net='sam2', out_size=1024, path_helper={'prefix': 'logs/REFUGE / medsam2_2024_08_13_07_04_47 ', 'ckpt_path': 'logs/REFUGE / medsam2_2024_08_13_07_04_47 /Model', 'log_path': 'logs/REFUGE / medsam2_2024_08_13_07_04_47 /Log', 'sample_path'：‘logs/ _medsam2_2024_08_13_07_04_47 /Samples’}, pretrain='MedSAM2_pretrain.pth', prompt='bbox', prompt_freq=2, sam_ckpt='./checkpoints/sam2_hiera_tiny.pt', sam_config='sam2_hiera_t', train_vis=False, val_freq=1, video_length=2, vis=True, weights=0)
> 
> Could you advise what could be the problem here ?你能告诉我这里有什么问题吗？
> 
> Thank you 谢谢你！

I encountered the same issue as you. Could you please share how you resolved it and give me some tips?  
Looking forward to your reply! Thank you!
